packages/lodestar/src/chain/reprocess.ts

```diff
@@ -0,0 +1,138 @@
+import {Slot, RootHex} from "@chainsafe/lodestar-types";
+import {IMetrics} from "../metrics";
+import {MapDef} from "../util/map";
+
+/**
+ * To prevent our node from having to reprocess while struggling to sync,
+ * we only want to reprocess attestations if block reaches our node before this time.
+ */
+export const REPROCESS_MIN_TIME_TO_NEXT_SLOT_SEC = 2;
+
+/**
+ * Reprocess status for metrics
+ */
+enum ReprocessStatus {
+  /**
+   * There are too many attestations that have unknown block root.
+   */
+  reached_limit = "reached_limit",
+  /**
+   * The awaiting attestation is pruned per clock slot.
+   */
+  expired = "expired",
+}
+
+type AwaitingAttestationPromise = {
+  resolve: (foundBlock: boolean) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  reject: (reason?: any) => void;
+  addedTimeMs: number;
+};
+
+// How many attestations (aggregate + unaggregate) we keep before new ones get dropped.
+const MAXIMUM_QUEUED_ATTESTATIONS = 16_384;
+
+type SlotRoot = {slot: Slot; root: RootHex};
+
+/**
+ * Some attestations may reach our node before the voted block, so we manage a cache to reprocess them
+ * when the block come.
+ * (n)                                               (n + 1)
+ *  |----------------|----------------|----------|------|
+ *                   |                |          |
+ *                  att           agg att        |
+ *                                              block
+ * Since the gossip handler has to return validation result to js-libp2p-gossipsub, this class should not
+ * reprocess attestations, it should control when the attestations are ready to reprocess instead.
+ */
+export class ReprocessController {
+  private readonly awaitingPromisesByRootBySlot: MapDef<Slot, MapDef<RootHex, AwaitingAttestationPromise[]>>;
```

Comment:
I wonder if we can make do with `MapDef<Slot, MapDef<RootHex, AwaitingAttestationPromise>>`
Reusing the same promise for attestations relying on the same block.
We would just be tracking the maximum wait time before resolve/reject per block.

Reply:
Good idea! Implemented :+1: